sockeye/inference.py

```diff
@@ -630,45 +635,86 @@ def make_input(sentence_id: int, sentence: str) -> TranslatorInput:
-                                    score=-np.inf)
+        # keep input ids to be able to concatenate translated sentence chunks back
+        chunk_markers = []
+        input_chunks = []
 
-        if len(trans_input.tokens) > self.max_input_length:
-            logger.debug("Input (%d) exceeds max input length (%d). Splitting into chunks of size %d.",
-                         len(trans_input.tokens), self.buckets_source[-1], self.max_input_length)
-            token_chunks = utils.chunks(trans_input.tokens, self.max_input_length)
-            translations = [self.translate_nd(*self._get_inference_input(tokens))
-                            for tokens in token_chunks]
-            translation = self._concat_translations(translations)
-            return self._make_result(trans_input, translation)
-        else:
-            return self._make_result(trans_input, self.translate_nd(*self._get_inference_input(trans_input.tokens)))
+        # prepare chunks, keeping track of empty or splitted inputs
+        for trans_input in trans_inputs:
+            if not trans_input.tokens:  # empty input?
+                chunk_markers.append(C.EMPTY_TRANSLATION_INPUT_ID)
+            elif len(trans_input.tokens) > self.max_input_length:  # oversized input?
+                logger.debug("Input %d has length (%d) that exceeds max input length (%d). Splitting into chunks of size %d.",
+                             trans_input.id, len(trans_input.tokens), self.buckets_source[-1], self.max_input_length)
+                token_chunks = list(utils.chunks(trans_input.tokens, self.max_input_length))
+                input_chunks.extend(token_chunks)
+                chunk_markers.extend([trans_input.id] * len(token_chunks))
+            else:
+                input_chunks.append(trans_input.tokens)
+                chunk_markers.append(trans_input.id)
+
+        # start translation in batches
+        translations = []  # type: List[Translation]
+        for i in range(0, len(input_chunks), self.batch_size):
+            logger.debug("Translating batch %d", i)
+            batch = input_chunks[i:i + self.batch_size]
+            # underfilled batch will be filled to a full batch size with copies of the 1st input
+            rest = self.batch_size - len(batch)
+            if rest > 0:
+                logger.debug("Extending the last batch to the full batch size (%d)", self.batch_size)
+                batch = batch + [batch[0]] * rest
+            # translate
+            translated_batch = self.translate_nd(*self._get_inference_input(batch))
+            # truncate to remove duplicated translations
+            if rest > 0:
+                translated_batch = translated_batch[:-rest]
+            translations.extend(translated_batch)
+
+        # concatenate translations corresponding to spans of chunk markers with the same input id
+        results = []  # type: List[TranslatorOutput]
+        start = 0
+        while start < len(chunk_markers):
```

Comment:
Just a suggestion but I think this could be simplified with `groupby`:
```
for translation_id, translations_and_markers in groupby(zip(translations, chunk_markers), key=[1]):
    translations_to_concat = [translation for translation, marker in translations_and_markers]
    if translation_id == C.EMPTY_TRANSLATION_INPUT_ID:
        for translation in translations_to_concat:
            results.append(TranslatorOutput(id=translation.id,
                                                     translation="",
                                                     tokens=[""],
                                                     attention_matrix=np.asarray([[0]]),
                                                     score=-np.inf))
    else:
        concatenated = translations_to_concat[0] if len(translations_to_concat) == 1 else \
                               self._concat_translations(translations_to_concat)
        results.append(self._make_result(trans_input, concatenated))
    
```

Reply:
Unfortunately doesn't work since chunks and translations have different sizes and are not aligned (there are no entries in translation for empty inputs, since we are not transalting them). But I like the groupby approach, - -will have to learn itertools better, maybe it can be fixed.

Reply:
ah yeah that's right, makes sense.